segfault when using additionalSecurityGroups with filters

/kind bug

**What steps did you take and what happened:**
I am modifying an existing `AWSMachine` and am adding this block here:
```yaml
  additionalSecurityGroups:
  - id: sg-0def541fd62cdc2ed
```

This works and attaches this SG to the machine as expected.

If I however provide this block here, I get a segfault:
```yaml
  additionalSecurityGroups:
  - filters:
    - name: group-id
      values:
      - sg-0def541fd62cdc2ed
```


**What did you expect to happen:**
Security group matching the selected filter to get attached to the instance.

**Anything else you would like to add:**
<details>
  <summary>Stacktrace</summary>
  
```
I0514 00:53:01.046787       1 awsmachine_controller.go:355] controllers/AWSMachine "msg"="Reconciling AWSMachine" "awsCluster"="sam-test1-a-us-east-1" "awsMachine"="sam-test1-a-us-east-1-node-f2dwk" "cluster"="sam-test1-a-us-east-1" "machine"="sam-test1-a-us-east-1-node-b759cfcdd-cvr6l" "namespace"="aws-908831091317"
E0514 00:53:01.053545       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 530 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1a33360, 0x2f81b00)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1a33360, 0x2f81b00)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).securityGroupsChanged(0xc000829f00, 0xc000b76240, 0xc000ca97c0, 0x2, 0x2, 0xc000c912f0, 0x1, 0x1, 0xc000b76210, 0xc000ca97a0, ...)
	/workspace/controllers/awsmachine_security_groups.go:77 +0x12e
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).ensureSecurityGroups(0xc000829f00, 0x209d5c0, 0xc000d684c8, 0xc00012d9f0, 0xc000c912f0, 0x1, 0x1, 0xc000b76210, 0x0, 0x204b380, ...)
	/workspace/controllers/awsmachine_security_groups.go:51 +0x112
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).reconcileNormal(0xc000829f00, 0x207ab00, 0xc000116058, 0xc00012d9f0, 0xc00048d100, 0x0, 0x0, 0x0, 0x20966a0)
	/workspace/controllers/awsmachine_controller.go:470 +0xaf1
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).Reconcile(0xc000829f00, 0xc0008b1030, 0x10, 0xc000704240, 0x20, 0xc0006bdc00, 0x0, 0x0, 0x0)
	/workspace/controllers/awsmachine_controller.go:172 +0xa99
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00023e3c0, 0x1abf180, 0xc0008701c0, 0xc000291000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00023e3c0, 0xc000291100)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc00023e3c0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0004365d0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0004365d0, 0x3b9aca00, 0x0, 0xc000291401, 0xc0000fc1e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc0004365d0, 0x3b9aca00, 0xc0000fc1e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:193 +0x328
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x17e641e]
```
</details>


**Environment:**

- Cluster-api-provider-aws version: `v0.5.3`
- Kubernetes version: (use `kubectl version`): `Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.4", GitCommit:"8d8aa39598534325ad77120c120a22b3a990b5ea", GitTreeState:"clean", BuildDate:"2020-03-12T20:55:23Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}`
- OS (e.g. from `/etc/os-release`): default ubuntu CAPA AMI

--- a/api/v1alpha3/awsmachine_webhook.go
+++ b/api/v1alpha3/awsmachine_webhook.go
@@ -53,6 +53,7 @@ func (r *AWSMachine) ValidateCreate() error {
 	allErrs = append(allErrs, r.validateRootVolume()...)
 	allErrs = append(allErrs, r.validateNonRootVolumes()...)
 	allErrs = append(allErrs, isValidSSHKey(r.Spec.SSHKeyName)...)
+	allErrs = append(allErrs, r.validateAdditionalSecurityGroups()...)
 
 	return aggregateObjErrors(r.GroupVersionKind().GroupKind(), r.Name, allErrs)
 }
@@ -182,4 +183,15 @@ func (r *AWSMachine) Default() {
 	if r.Spec.CloudInit.SecureSecretsBackend == "" {
 		r.Spec.CloudInit.SecureSecretsBackend = SecretBackendSecretsManager
 	}
+}
+
+func (r *AWSMachine) validateAdditionalSecurityGroups() field.ErrorList {
+	var allErrs field.ErrorList
+
+	for _, additionalSecurityGroups := range r.Spec.AdditionalSecurityGroups {
+		if len(additionalSecurityGroups.Filters) > 0 {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec.additionalSecurityGroups"), "filters are not implemented for security groups and will be removed in a future release"))
+		}
+	}
+	return allErrs
 }

--- a/api/v1alpha3/awsmachine_webhook_test.go
+++ b/api/v1alpha3/awsmachine_webhook_test.go
@@ -137,6 +137,24 @@ func TestAWSMachine_ValidateCreate(t *testing.T) {
 			},
 			wantErr: false,
 		},
+		{
+			name: "additional security groups should not have filters",
+			machine: &AWSMachine{
+				Spec: AWSMachineSpec{
+					AdditionalSecurityGroups: []AWSResourceReference{
+						{
+							Filters: []Filter{
+								{
+									Name:   "example-name",
+									Values: []string{"example-value"},
+								},
+							},
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

